crypto/evp/p_lib.c

```diff
@@ -732,6 +736,110 @@ int EVP_PKEY_base_id(const EVP_PKEY *pkey)
     return EVP_PKEY_type(pkey->type);
 }
 
+int EVP_PKEY_is_a(const EVP_PKEY *pkey, const char *name)
+{
+#ifndef FIPS_MODE
+    if (pkey->keymgmt == NULL)
+        return EVP_PKEY_type(pkey->type) == EVP_PKEY_type(OBJ_sn2nid(name));
+#endif
+    return EVP_KEYMGMT_is_a(pkey->keymgmt, name);
+}
+
+int EVP_PKEY_can_sign(const EVP_PKEY *pkey)
+{
+    if (pkey->keymgmt == NULL) {
+        switch (EVP_PKEY_base_id(pkey)) {
+        case EVP_PKEY_RSA:
+            return 1;
+#ifndef OPENSSL_NO_DSA
+        case EVP_PKEY_DSA:
+            return 1;
+#endif
+#ifndef OPENSSL_NO_EC
+        case EVP_PKEY_ED25519:
+        case EVP_PKEY_ED448:
+        case EVP_PKEY_EC:        /* Including SM2 */
+            return EC_KEY_can_sign(pkey->pkey.ec);
+#endif
+        default:
+            return 0;
+        }
+    } else {
+        const OSSL_PROVIDER *prov = EVP_KEYMGMT_provider(pkey->keymgmt);
+        OPENSSL_CTX *libctx = ossl_provider_library_context(prov);
+        const char *supported_sig = NULL;
+        EVP_SIGNATURE *signature = NULL;
+
+        supported_sig =
+            pkey->keymgmt->query_operation_name != NULL
+            ? pkey->keymgmt->query_operation_name(OSSL_OP_SIGNATURE)
+            : evp_first_name(prov, pkey->keymgmt->name_id);
+        signature = EVP_SIGNATURE_fetch(libctx, supported_sig, NULL);
+        EVP_SIGNATURE_free(signature);
+        return signature != NULL;
```

Comment:
A pointer is just a value just like any other value.  That value won't "magically" change because you freed the memory area its value points at.  The value in itself is perfectly usable as long as it's not dereferenced.

But, I can waste another variable to hold the truth value if you prefer...

Reply:
FWIW, I agree that looking at the pointer after it's free'd, while technically legal, is confusing.

Reply:
I think that *technically* it's not actually legal.
It helps to recall that the C rules for pointers have their origins in the era of segmented architectures, where even trying to load a pointer value (not dereference it) for which the corresponding segment was not valid anymore, could fault.

Reply:
Huh!  I didn't know that.

Reply:
You could nicely do 
```
if (signature != NULL) {
    EVP_SIGNATURE_free(signature); 
    return 1; 
} return 0;
```

Reply:
I believe Ben is wrong; "free(p); look at p but not *p" is legal.

Reply:
See n1256.pdf appendix J.2, stanzas 8 and 9:

```
— An object is referred to outside of its lifetime (6.2.4).
— The value of a pointer to an object whose lifetime has ended is used (6.2.4).
```
The first one refers to dereferencing the pointer, and the second refers to only the value of the (undereferenced) pointer.

Reply:
I stand corrected, thanks.